compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -292,23 +292,58 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def TypeDef(sym: TypeSymbol)(using Context): TypeDef =
     ta.assignType(untpd.TypeDef(sym.name, TypeTree(sym.info)), sym)
 
-  def ClassDef(cls: ClassSymbol, constr: DefDef, body: List[Tree], superArgs: List[Tree] = Nil)(using Context): TypeDef = {
+  def ClassDef(cls: ClassSymbol, constrDef: DefDef, body: List[Tree])(using Context): TypeDef = {
     val firstParent :: otherParents = cls.info.parents: @unchecked
     val superRef =
       if (cls.is(Trait)) TypeTree(firstParent)
       else {
-        def isApplicable(ctpe: Type): Boolean = ctpe match {
-          case ctpe: PolyType =>
-            isApplicable(ctpe.instantiate(firstParent.argTypes))
-          case ctpe: MethodType =>
-            (superArgs corresponds ctpe.paramInfos)(_.tpe <:< _)
+        def fromRepeated(tpe: Type): Type = tpe match {
+          case AnnotatedType(AppliedType(base, List(tparam)), annot) if annot.symbol == defn.RepeatedAnnot =>
+            tparam
           case _ =>
-            false
+            NoType
         }
-        val constr = firstParent.decl(nme.CONSTRUCTOR).suchThat(constr => isApplicable(constr.info))
-        New(firstParent, constr.symbol.asTerm, superArgs)
+
+        def canConstructMethod(constrSym: Symbol): Boolean = constrSym.info match {
+          case ctpe: MethodOrPoly =>
+            constrSym.paramSymss.flatten.filter(_.isTerm).forall { paramSym =>
+              paramSym.flags.is(Flags.HasDefault) || (fromRepeated(paramSym.info) != NoType)
+            }
+          case _ => false
+        }
+
+        def constructParent(constr: Tree): Tree = {
+          constr.symbol.paramSymss.filter(!_.exists(_.isType)).foldLeft(constr) { (acc, paramList) =>
+            acc.appliedToTermArgs(paramList.zipWithIndex.collect {
+              case (paramSym, idx) if paramSym.flags.is(Flags.HasDefault) =>
+                typer.Applications.defaultArgument(acc, idx, false)
```

Comment:
This is very unusual and dangerous. For the sake of modularity, we should not refer to typer internals from transform phases. TypeAssigner is fine, but typer should be off limits.